docs/core/upgrade.md

```diff
@@ -0,0 +1,108 @@
+
+### Version Map
+The version map is a mapping of module names to consensus versions. The map is persisted to x/upgrade's state for use during in-place migrations. When migrations finish, the updated version map is persisted to state. 
+
+## Upgrade Handlers
+
+Upgrades utilize an `UpgradeHandler` to facilitate migrations. `UpgradeHandler`s are functions implemented by the app developer that conform to the following function signature. These functions retrieve the `VersionMap` from x/upgrade's state, and return the new `VersionMap` to be stored in x/upgrade after the upgrade. The diff between the two `VersionMap`s determines which modules need upgrading.
+
+```golang
+type UpgradeHandler func(ctx sdk.Context, plan Plan, fromVM VersionMap) (VersionMap, error)
+```
+
+Inside these functions, you should perform any upgrade logic you wish to include in the provided `plan`. All upgrade handler functions should end with the following line of code:
+
+```golang
+  return app.mm.RunMigrations(ctx, cfg, fromVM)
+```
+
+## Running Migrations
+
+Migrations are run inside of an `UpgradeHandler` via `app.mm.RunMigrations(ctx, cfg, vm)`. As described above, `UpgradeHandler`s are functions which describe the functionality to occur during an upgrade. The `RunMigration` function will loop through the `VersionMap` argument, and run the migration scripts for any versions that are less than the new binary's app module versions. Once the migrations are finished, a new `VersionMap` will be returned to persist the upgraded module versions to state.
+
+```golang
+cfg := module.NewConfigurator(...)
+app.UpgradeKeeper.SetUpgradeHandler("my-plan", func(ctx sdk.Context, plan upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
+
+    // ...
+    // do upgrade logic
+    // ...
+
+    // RunMigrations returns the VersionMap
+    // with the updated module ConsensusVersions
+    return app.mm.RunMigrations(ctx, vm)
+})
+```
+
+To learn more about configuring migration scripts for your modules, refer to this [guide](../building-modules/upgrade.md).
+
+## Adding New Modules In Upgrades
+
+Entirely new modules can be introduced to the application during an upgrade. The SDK recognizes new modules during upgrades because their consensus version in the `fromVM` `VersionMap` is 0. In this case, `RunMigrations` will call the corresponding module's `InitGenesis` function to setup its initial state. This can be skipped if the module does not require any inital state. Otherwise, it is important to implement `InitGenesis` for new modules to successfully upgrade your application without error.
```

Comment:
```suggestion
You can introduce entirely new modules to the application during an upgrade. New modules are recognized because their consensus version in the `fromVM` `VersionMap` is 0. In this case, `RunMigrations` calls the  `InitGenesis` function from the corresponding module to set up its initial state. If the module requires an initial state, run `InitGenesis` to ensure new modules can successfully upgrade your application without error. 
```